## 2.4.0 - 2024-06-25
[GRD-797](https://jira.oicr.on.ca/browse/GRD-797) - add vidarr labels to outputs (changes to medata only)
## 2.3.3 - 2024-02-12
- [GRD-730](https://jira.oicr.on.ca/browse/GRD-730) Adding minMemory parameter to RAM-scaling tasks
## 2.3.2 - 2023-12-21
- [GRD-730](https://jira.oicr.on.ca/browse/GRD-730) Adding RAM scaling by chromosome size (performance update)
## 2.3.1 - 2023-09-20
- Update mouse genome build and reference-specific modules to mm39 (from mm10; Jira Ticket: GBS-4248)
## 2.3.0 - 2023-07-05
- Assembly-specific codes in olive moved to wdl
## 2.2.1 - 2023-05-25
- add tumor and normal names as parameters
## 2.2.0 - 2023-03-01
- move version 2.1.8 to version 2.2.0
## 2.1.8 - 2023-02-14
- add an option vepStats
- update the vcf2maf module to 1.6.21b
## 2.1.7 - 2022-09-22
- migrate 2.1.6 to Vidarr
- fixing getSampleNames
## 2.1.6 - 2022-04-13
- update to support vep/105
## 2.1.51 - 2023-03-01
- small fix to the version 2.1.5, fixing getSampleNames
## 2.1.5 - 2022-06-28
- changing targetBed file type to String
## 2.1.4 - 2021-06-01
- migrate to Vidarr
## 2.1.3 - 2021-02-08
- addition of retainInfoProvided and updateTagValue options
## 2.1.2 - 2020-11-10
- supporting non-human species
## 2.1.1 - 2020-09-23
- fixing chromosome order
## 2.1 - 2020-09-11
- run vep and vcf2maf in parallel
## 2.0.2 - 2020-07-02
- supporting normal samples without groupID
## 2.0.1 - 2020-06-08
- fixing the header for tumor-only vcf
## 2.0 - 2020-05-07
- WDL workflow for VEP
